packages/core/src/shared/utilities/processUtils.ts

```diff
@@ -61,14 +63,144 @@ export interface ChildProcessResult {
 
 export const eof = Symbol('EOF')
 
+export interface ProcessStats {
+    memory: number
+    cpu: number
+}
+export class ChildProcessTracker {
+    static readonly pollingInterval: number = 10000 // Check usage every 10 seconds
+    static readonly thresholds: ProcessStats = {
+        memory: 100 * 1024 * 1024, // 100 MB
+        cpu: 50,
+    }
+    static readonly logger = getLogger('childProcess')
+    #processByPid: Map<number, ChildProcess> = new Map<number, ChildProcess>()
+    #pids: PollingSet<number>
+
+    public constructor() {
+        this.#pids = new PollingSet(ChildProcessTracker.pollingInterval, () => this.monitor())
+    }
+
+    private cleanUp() {
+        const terminatedProcesses = Array.from(this.#pids.values()).filter(
+            (pid: number) => this.#processByPid.get(pid)?.stopped
+        )
+        for (const pid of terminatedProcesses) {
+            this.delete(pid)
+        }
+    }
+
+    private async monitor() {
+        this.cleanUp()
+        ChildProcessTracker.logger.debug(`Active running processes size: ${this.#pids.size}`)
+
+        for (const pid of this.#pids.values()) {
+            await this.checkProcessUsage(pid)
+        }
+    }
+
+    private async checkProcessUsage(pid: number): Promise<void> {
+        if (!this.#pids.has(pid)) {
+            ChildProcessTracker.logger.warn(`Missing process with id ${pid}`)
+            return
+        }
+        const stats = this.getUsage(pid)
+        if (stats) {
+            ChildProcessTracker.logger.debug(`Process ${pid} usage: %O`, stats)
+            if (stats.memory > ChildProcessTracker.thresholds.memory) {
+                ChildProcessTracker.logger.warn(`Process ${pid} exceeded memory threshold: ${stats.memory}`)
+            }
+            if (stats.cpu > ChildProcessTracker.thresholds.cpu) {
+                ChildProcessTracker.logger.warn(`Process ${pid} exceeded cpu threshold: ${stats.cpu}`)
+            }
+        }
+    }
+
+    public add(childProcess: ChildProcess) {
+        const pid = childProcess.pid()
+        this.#processByPid.set(pid, childProcess)
+        this.#pids.start(pid)
+    }
+
+    public delete(childProcessId: number) {
+        this.#processByPid.delete(childProcessId)
+        this.#pids.delete(childProcessId)
+    }
+
+    public get size() {
+        return this.#pids.size
+    }
+
+    public has(childProcess: ChildProcess) {
+        return this.#pids.has(childProcess.pid())
+    }
+
+    public clear() {
+        for (const childProcess of this.#processByPid.values()) {
+            childProcess.stop(true)
+        }
+        this.#pids.clear()
+        this.#processByPid.clear()
+    }
+
+    public getUsage(pid: number): ProcessStats {
+        try {
+            // isWin() leads to circular dependency.
+            return process.platform === 'win32' ? getWindowsUsage() : getUnixUsage()
+        } catch (e) {
+            ChildProcessTracker.logger.warn(`Failed to get process stats for ${pid}: ${e}`)
```

Comment:
If it ever fails we should probably not attempt on future invocations?

Reply:
Not really sure what could cause it to fail here. Do you think there is a risk in continually retrying? Maybe potential log spam? Could this be combatted by setting the polling interval longer?

Reply:
> Not really sure what could cause it to fail here.

Permissions issues usually. 

> Maybe potential log spam?

Yeah that's at least my first thought. When a system is in a bad state, it can make things worse if logs "cascade" with a deluge of redundant warnings.